### ISSUE TITLE
Slicer crashes on Windows when volume with NaN voxel value is loaded

### ISSUE THREAD

**Issue by lassoan** [Slicer/Slicer#7737]
## Summary

Slicer crashes when attempting to load a volume that has some NaN-valued voxels.

The crash is reproducible with `vtkSlicerVolumesLogicTest1_TestNAN` test, which fails on Windows since February 2024 due to a crash in VTK's `vtkImageHistogramExecute` function in `vtkImageHistogram.cxx`. This code should return a value between xmin and xmax if x is nan:

```
x = (x > xmin ? x : xmin);
x = (x < xmax ? x : xmax);
```

However, x remains nan, therefore `xi` bin index is set to a negative number, and so `outPtr[xi]++;` causes access violation. x remains nan due to an MSVC compiler bug: https://stackoverflow.com/questions/77799136/msvc-2022-wrong-optimization-of-maxnan-somenumber

## Steps to reproduce

Load [testNANInVolume.nrrd](https://github.com/Slicer/SlicerTestingData/releases/download/SHA256/b542f49c4aa056f55e01d12e94c9991e8fc335f5204b380e6cc70a02d206e7ad) => Slicer crashes

Run vtkSlicerVolumesLogicTest1_TestNAN test => fails (for example: https://slicer.cdash.org/tests/27309466)

## Environment
- Slicer version: Slicer-5.7.0-204-04-15
- Operating system: Windows, MSVC toolset 19.39.33523

**Comment by lassoan** [comment]
The crash could be fixed in VTK by adding an explicit `if (!std::isnan(x))` check, but maybe it is OK to just wait for a compiler update, as NaN-valued volumes are not very common.

```
//------------------------------------------------------------------------------
template <class T>
void vtkImageHistogramExecute(vtkImageHistogram* self, vtkImageData* inData,
  vtkImageStencilData* stencil, T* inPtr, int extent[6], vtkIdType* outPtr, int binRange[2],
  double o, double s, int component, int threadId)
{
  vtkImageStencilIterator<T> inIter(inData, stencil, extent, ((threadId == 0) ? self : nullptr));

  // set up components
  int nc = inData->GetNumberOfScalarComponents();
  int c = component;
  if (c < 0)
  {
    nc = 1;
    c = 0;
  }

  // compute shift/scale values for fast bin computation
  double xmin = binRange[0];
  double xmax = binRange[1];
  double xshift = -o;
  double xscale = 1.0 / s;

  // iterate over all spans in the stencil
  while (!inIter.IsAtEnd())
  {
    if (inIter.IsInStencil())
    {
      inPtr = inIter.BeginSpan();
      T* inPtrEnd = inIter.EndSpan();

      // iterate over all voxels in the span
      if (inPtr != inPtrEnd)
      {
        int n = static_cast<int>((inPtrEnd - inPtr) / nc);
        inPtr += c;
        do
        {
          double x = *inPtr;

          if (!std::isnan(x))
          {
            x += xshift;
            x *= xscale;

            // x may be NaN but we don't want xmin and xmax
            // to be ever end up being set to NaN.
            // Therefore x must be used as a second operand
            // of the ternary operator (because result of comparing
            // NaN to any other value is always false).
            x = (x > xmin ? x : xmin);
            x = (x < xmax ? x : xmax);

            int xi = static_cast<int>(x + 0.5);

            outPtr[xi]++;
          }

          inPtr += nc;
        } while (--n);
      }
    }
    inIter.NextSpan();
  }
}
```

**Comment by pieper** [comment]
If this crashes with an officially supported compiler then IMO it would be better to fix it rather than waiting for a compiler fix.  Research codes sometimes generate NaNs.  But it's not a super high priority for me.

**Comment by lassoan** [comment]
Yes, this crash happens with the latest MSVC compiler. A new version is expected within a few months, which may or may not have a fix for this.

**Comment by jcfr** [comment]
As of Sept 17th 2024, test is still failing on the "factory":

```
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.39.33523.0
-- The CXX compiler identification is MSVC 19.39.33523.0
```

See https://slicer.cdash.org/tests/28285893

**Comment by lassoan** [comment]
I've updated to latest `Visual Studio 2022 (v143)` platform toolset, which contained `Microsoft (R) C/C++ Optimizing Compiler Version 19.41.34120 for x64` and with this compiler the test was **successful**.

We should upgrade the compiler on the factory build to get a fix for this error.

**Comment by jamesobutler** [comment]
Going to 19.41 will likely still run into the issue of https://github.com/Slicer/Slicer/issues/7847 if the computer running Slicer does not have the latest MSVC redistributable. Unfortunately I was unable to get things working with the compile flag change.

**Comment by lassoan** [comment]
I did not change any compilation flags, just updated the compiler. The bug may be in code that the compiler generated, not necessarily in runtime libraries. Also, as far as I remember, we include msvc runtime dlls in the Slicer package.

**Comment by jamesobutler** [comment]
Just to clarify, updating to 19.41 may fix this issue, but it will likely still introduce the issue that I reported at #7847 which started to happen with MSVC 19.40 and would become a more common issue if the Slicer factory built installed package was using version 19.40+.

**Comment by jcfr** [comment]
@jamesobutler @lassoan @sjh26 

Now that we have a fix for #7847 available for review in #8149, I would like to move forward with updating the version of Visual Studio used on our windows factory.

We will then be able to use the updated compiler for both the `Preview` and latest `Stable` release.

**Comment by jcfr** [comment]
I will shortly announce we are initiating the release process of Slicer 5.8.

Once we move forward with the announcement, the build of stable extensions will be disabled.

In the meantime, I would like to look into addressing few additional issues (see updated [5.7 milestone](https://github.com/Slicer/Slicer/milestone/10)

**Comment by jcfr** [comment]
Visual Studio has been updated from `17.9.6` to `17.12.4`

### MSVC Compiler Version

Before:

```
[...]
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.39.33523.0
-- The CXX compiler identification is MSVC 19.39.33523.0
[...]
```

After:

```
[...]
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.42.34436.0
-- The CXX compiler identification is MSVC 19.42.34436.0
[...]
```


See https://discourse.slicer.org/t/temporary-disabling-of-stable-extension-builds-in-preparation-for-slicer-5-8-release-visual-studio-update/41207/5?u=jcfr

**Comment by jcfr** [comment]
Following the update, the test `vtkSlicerVolumesLogicTest1_TestNAN` is now passing on Windows:

|  https://slicer.cdash.org/tests/29338405?graph=status |
|--|
| ![Image](https://github.com/user-attachments/assets/6fc46615-cbd9-4543-bdbe-0e67201154b0) |
